datafusion/core/src/physical_optimizer/enforcement.rs

```diff
@@ -0,0 +1,1739 @@
+                } else {
+                    plan
+                }
+            }
+            _ => plan,
+        }
+    } else if let Some(SortMergeJoinExec {
+        left,
+        right,
+        on,
+        join_type,
+        sort_options,
+        null_equals_null,
+        ..
+    }) = plan_any.downcast_ref::<SortMergeJoinExec>()
+    {
+        let join_key_pairs = extract_join_keys(on);
+        if let Some((
+            JoinKeyPairs {
+                left_keys,
+                right_keys,
+            },
+            new_positions,
+        )) = reorder_current_join_keys(
+            join_key_pairs,
+            Some(left.output_partitioning()),
+            Some(right.output_partitioning()),
+            &plan.equivalence_properties(),
+        ) {
+            if !new_positions.is_empty() {
+                let new_join_on = new_join_conditions(&left_keys, &right_keys);
+                let mut new_sort_options = vec![];
+                for idx in 0..sort_options.len() {
+                    new_sort_options.push(sort_options[new_positions[idx]])
+                }
+                Arc::new(
+                    SortMergeJoinExec::try_new(
+                        left.clone(),
+                        right.clone(),
+                        new_join_on,
+                        *join_type,
+                        new_sort_options,
+                        *null_equals_null,
+                    )
+                    .unwrap(),
+                )
+            } else {
+                plan
+            }
+        } else {
+            plan
+        }
+    } else {
+        plan
+    }
+}
+
+/// Reorder the current join keys ordering based on either left partition or right partition.
+fn reorder_current_join_keys(
+    join_keys: JoinKeyPairs,
+    left_partition: Option<Partitioning>,
+    right_partition: Option<Partitioning>,
+    equivalence_properties: &EquivalenceProperties,
+) -> Option<(JoinKeyPairs, Vec<usize>)> {
+    match (left_partition.clone(), right_partition.clone()) {
+        (Some(Partitioning::Hash(left_exprs, _)), _) => {
+            try_reorder(join_keys.clone(), left_exprs, equivalence_properties).or_else(
+                || {
+                    reorder_current_join_keys(
+                        join_keys,
+                        None,
+                        right_partition,
+                        equivalence_properties,
+                    )
+                },
+            )
+        }
+        (_, Some(Partitioning::Hash(right_exprs, _))) => {
+            try_reorder(join_keys.clone(), right_exprs, equivalence_properties).or_else(
+                || {
+                    reorder_current_join_keys(
+                        join_keys,
+                        left_partition,
+                        None,
+                        equivalence_properties,
+                    )
+                },
+            )
+        }
+        _ => None,
+    }
+}
+
+fn try_reorder(
```

Comment:
It's better to add some comments here for the parameters.

- For the top-down case, the `join_keys` are using the its children's schema; while the `expected` and `equivalence_properties` are using the current operator's schema.
- For the bottom-up case, the `join_keys` and `expected` are using the its children's schema; while the `equivalence_properties` are using the current operator's schema. 

Is it possible to make the schema the same for all of the parameters? Otherwise, for the right side of the bottom-up case, it may miss some reorder chance.

Reply:
Yes, valid points.

Reply:
Now, in the latest code, for the bottom-up case, the join_keys, expected and equivalence_properties are all using the its children's schema.